File: python-backend/askquestion.py
from fastapi import APIRouter, Request
from fastapi.responses import JSONResponse
import json
from openai import OpenAI
from util import generateToken
import os

generateToken()
router = APIRouter()

@router.get("/ask-question")
async def generate_question(topic: str):
    print("Generating questions for topic: ", topic)
    # Challenge 1.a - Write the prompt
   # Challenge 1.a - Write the prompt
    prompt = f"""
    You are an expert in {topic}. You will provide an interview question.
    Ensure that only one question is shown.
    Ensure that the question is prefixed with "Question:".
    Ensure that the returned question doesn't have an answer included.
    """
    header_name = os.getenv('GATEWAY_HEADER_NAME')
    header_value = os.getenv('GATEWAY_HEADER_VALUE')
    headers = {
        header_name: header_value
    }
    client = OpenAI(default_headers=headers)

    messages = [{"role": "user", "content": prompt}]
    # Challenge 1.b - Call OpenAI API to generate questions
    response = client.chat.completions.create(
        model="gpt-4o-2024-08-06",
        messages=messages,
    )
  
    # Extract the generated text
    questions_text = response.choices[0].message.content

    # Return the questions as a JSON response
    return JSONResponse(content={"question": questions_text})

@router.post("/question-feedback")
async def submit_descriptive_questions(request: Request):
    # Request parameter includes both question and answer provided by user.
    data = await request.json()
    question = data['question']['question']
    answer = data['question']['CandidateAnswer']
    
    #Challenge 2.a - Write Prompt to evaluate the question and answer
    #Use question and answer variable to generate the prompt
    prompt = f""" 
    
                      """

    # OpenAI Call to generate feedback
    header_name = os.getenv('GATEWAY_HEADER_NAME')
    header_value = os.getenv('GATEWAY_HEADER_VALUE')
    headers = {
        header_name: header_value
    }
    client = OpenAI(default_headers=headers)

    #Challenge 2.b - Call the OpenAI API and get the resopnse
    response = {}

    # Extract the generated text
    feedback_text = response.choices[0].message.content

    # Return the feedback as a JSON response
    return JSONResponse(content={"feedback": feedback_text})

File: nodejs-backend/routes/askquestion.js
const express = require('express');
const router = express.Router();
const { generateToken } = require('../util');
const { OpenAI } = require('openai');

(async () => {
  await generateToken();
})();

// Endpoint to generate a question
router.get('/ask-question', async (req, res) => {
  const topic = req.query.topic;

  // Challenge 1.a - Write the prompt
  const prompt = ``;

  const header_name = process.env.GATEWAY_HEADER_NAME
  const header_value = process.env.GATEWAY_HEADER_VALUE
  const headers = {
    header_name: header_value,
  };
  const client = new OpenAI(headers);

  // Challenge 1.b - Call OpenAI API to generate questions
  const response = {}

  // Extract the generated text
  const questions_text = response.choices[0].message.content;

  // Return the questions as a JSON response
  res.json({ question: questions_text });
});

// Endpoint to submit descriptive questions
router.post('/question-feedback', async (req, res) => {
  // Request parameter includes both question and answer provided by user.
  const { question, CandidateAnswer: answer } = req.body.question;

  // Challenge 2.a - Write Prompt to evaluate the question and answer
  // Use question and answer variable to generate the prompt
  const prompt = ``;

  // OpenAI Call to generate feedback
  const header_name = process.env.GATEWAY_HEADER_NAME
  const header_value = process.env.GATEWAY_HEADER_VALUE
  const headers = {
    header_name: header_value,
  };
  const client = new OpenAI(headers);

  // Challenge 2.b - Call the OpenAI API and get the response
  const response = {

  }
  // Extract the generated text
  const feedback_text = response.choices[0].message.content;

  // Return the feedback as a JSON response
  res.json({ feedback: feedback_text });
});

module.exports = router;
